crates/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -707,11 +707,12 @@ pub fn comparator_operand_type_rules(
             // We could check if all elements of all arrays are comptime but I am lazy
             Ok(Bool(Comptime::No(Some(op.location.span))))
         }
-        (lhs, rhs) => {
-            if lhs == rhs {
+        (NamedGeneric(binding_a, name_a), NamedGeneric(binding_b, name_b)) => {
```

Comment:
NamedGenerics can refer to any type, most of which cannot be compared

Reply:
Yes, but you do not know at this stage if they can or not.

Reply:
Generic types always cannot be compared, they have no behaviors that are valid for all types. Allowing this and dynamically failing would be having a template system instead of a generics system which was not a design decision we chose for noir

Reply:
This will not be a dynamic failing, we should be able to fail after monomorphisation.

Reply:
Yes, this is what I meant: whether we fail in the type checker or after monomorphising possibly several functions. Doing the later would lead to us having to issue full call stacks for these type errors since they'd be caught at the callsite in a library when we eventually have a `fn1 < fn2` rather than where the user called into a library function originally.

Reply:
it is better to fail later rather than restricting the features of the language.

Reply:
We are not restricting language features, it is just that generics are still relying on higher order functions for most features since they'd rely on traits otherwise where we would similarly need an `Ord` trait bound. We should implement these needed features rather than leaving type checking for the ssa pass.

Reply:
Not being able to sort arrays in a generic way is a restriction! We can add the restriction once we have traits but for now we should allow this. It will break backward compatibility but it is understandable for this feature.

Reply:
It fails with this as we do not support array of structs. So a user will likely get errors before being able to (try to) sort (although we can fix those).
`thread 'main' panicked at 'not implemented: Conversion to ObjectType is unimplemented for type Tuple([Field])', crates/noirc_evaluator/src/ssa/node.rs:224:17`

Reply:
Structs or tuples don't need to be behind an array to be compared. I tried it locally and the error I get is:
```
thread 'main' panicked at 'Tried to unwrap a struct into a single value', crates/noirc_evaluator/src/ssa/code_gen.rs:43:32
stack backtrace:
   0: std::panicking::begin_panic
             at /rustc/e7119a0300b87a3d670408ee8e847c6821b3ae80/library/std/src/panicking.rs:588:12
   1: noirc_evaluator::ssa::code_gen::Value::unwrap_id
   ... full backtrace ...
```
From the code
```rs
fn compare<T>(a: T, b: T) -> bool {
    a < b
}

fn main(x: Field, y: Field) {
    let pair = (x, y);
    constrain compare(pair, pair);
}
```

Reply:
This error can be easily improved, I just made a commit to show this. The message can be improved and it could distinguish between () and struct/tuple.